src/librustc/lint/builtin.rs

```diff
@@ -352,6 +352,19 @@ declare_lint! {
     "outlives requirements can be inferred"
 }
 
+declare_lint! {
+    pub DEPRECATED_IN_FUTURE,
+    Allow,
+    "detects use of items that will be deprecated in a future version",
+    report_in_external_macro: true
+}
+
+declare_lint! {
+    pub DUPLICATE_AUTO_TRAITS_IN_BOUNDS,
```

Comment:
I think we should just call this `DUPLICATE_BOUNDS` In particular, I think we could easily extend this in the future to non-auto-traits as well as to other sorts of bounds (e.g., `where Self: 'static, Self: 'static` etc).

Reply:
I agree we should keep this general for any sort of bound... However, `Foo: Bar` and `Bar + Baz` cannot both be bounds at the same time; I call the former a *constraint* and the latter a *bound*; Thus, `Self: 'static, Self: 'static` and `Self: 'static + 'static` are not the same thing, the first is duplicate constraints and the latter duplicate bounds. (tho you can desugar the latter to the former)

Reply:
Interesting. I'm not sure I see what distinction you are drawing, precisely. Is if the difference between `Foo: Bar` (where the self type is stated) and just `Bar` (which is missing the self type), or is the different between `Foo: Bar` and `Foo: Bar + Baz`? 

In the compiler / in the past, I think we draw the distinction between a *where clause* (which includes the self type, so e.g. `Foo: Bar`) and a *bound* (which is just the `Bar` part).

I think that the term "bound" and "constraint" don't make it particularly clear what we are referring to. They seem basically like synonyms to me. I wonder if we can use modifiers instead?

- `Foo: Bar` -- a **bound**
- `Foo: Bar + Baz` -- a **compound bound**
- `Bar` -- an .. incomplete bound? =)

OTOH I'm not sure if it's really important to have a term for `Bar` without the self type -- though it obviously comes up for trait objects, at least *grammatically*. Still I feel like we could get by just using one term.

In general, I like the term "where clause" because it seems to be clearly linked to Rust syntax, and hence the most obvious. But I fear that people will assume (reasonably) that a "where clause" *must* be introduced with `where`, and hence may be confused if we refer to the `Bar` in `fn foo<T: Bar>()` or `dyn Bar` as a "where clause".

Reply:
> Interesting. I'm not sure I see what distinction you are drawing, precisely. Is if the difference between `Foo: Bar` (where the self type is stated) and just `Bar` (which is missing the self type), or is the difference between `Foo: Bar` and `Foo: Bar + Baz`?

The difference between former; i.e. `Foo: Bar + 'a` vs. `Bar + 'a`.
In my model, I think of Rust roughly (forgetting about `for<'a>` and such...) as having:
+ base kinds `k0 ::= type | lifetime;`
+ bounds as "incomplete constraints": `bound = k0 -> constraint` ("1-argument curried constraint")
+ `+` as a binary operator `bound, bound -> bound` (tho in reality its a list)
+ axiom sets: `where [constraint]*`

> In the compiler / in the past, I think we draw the distinction between a _where clause_ (which includes the self type, so e.g. `Foo: Bar`) and a _bound_ (which is just the `Bar` part).

In Wadler's terminology (as per the paper [Quantified Class Constraints](https://homepages.inf.ed.ac.uk/wadler/papers/quantcc/quantcc.pdf)) a `where` clause would correspond to an *"axiom set"* (denoted by `A` in the paper) and an axiom set consists of a list of *"constraints"*.

> I think that the term "bound" and "constraint" don't make it particularly clear what we are referring to.

ISTM that this is mainly due to using "bound" both for `Foo: Bar` and `Bar` which is confusing. I find that we primarily use "bound" for `Bar` right now. For the reference we need clear terminology, and since Wadler came up with type classes, my preference is for using Wadler's "constraint" terminology... ;)

> They seem basically like synonyms to me. I wonder if we can use modifiers instead?
> 
> * `Foo: Bar` -- a **bound**
> * `Foo: Bar + Baz` -- a **compound bound**
> * `Bar` -- an .. incomplete bound? =)

I'd categorize as this instead:
> * `where Foo: Bar, Baz: Quux` -- a **where clause** (or axiom set, but we don't need to call it that...)
> * `Foo: Bar` -- a **constraint**
> * `for<'a> Foo<'a>: Bar` -- still a constraint, but generic/universally quantified over lifetimes
> * `Bar + Baz` -- a **compound bound**
> * `Bar` -- a **bound**
> * `for<'a> Bar<'a>` -- still a bound, but generic over lifetimes

where bounds are constraint-constructors (but we don't need to call it that in user documentation...)

> 
> OTOH I'm not sure if it's really important to have a term for `Bar` without the self type -- though it obviously comes up for trait objects, at least _grammatically_. Still I feel like we could get by just using one term.

I feel like at least for the sake of not confusing ourselves (which I find happens often) we need 2 terms... trait objects are a good reason distinguish; we can think of `dyn` as an operator `bound -> type`.

> In general, I like the term "where clause" because it seems to be clearly linked to Rust syntax, and hence the most obvious. But I fear that people will assume (reasonably) that a "where clause" _must_ be introduced with `where`, and hence may be confused if we refer to the `Bar` in `fn foo<T: Bar>()` or `dyn Bar` as a "where clause".

`<T: Bar>` is a bit tricky here since it's both quantification and an axiom set at the same time... I think we should probably leave "where clause" to the syntactic rather than semantic construct...